prow/spyglass/gcsartifact.go

```diff
@@ -0,0 +1,240 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+// GCSArtifact represents some output of a prow job stored in GCS
+type GCSArtifact struct {
+	// The handle of the object in GCS
+	handle artifactHandle
+
+	// The link to the Artifact in GCS
+	link string
+
+	// The path of the Artifact within the job
+	path string
+
+	// sizeLimit is the max size to read before failing
+	sizeLimit int64
+
+	// ctx provides context for cancellation and timeout. Embedded in struct to preserve
+	// conformance with io.ReaderAt
+	ctx context.Context
+}
+
+type artifactReader interface {
```

Comment:
Isn't this just https://golang.org/pkg/io/#ReadCloser?

Reply:
Yes but someday we may use other methods from `*storage.Reader` which is what this was meant to embed

Reply:
Sure,

In general I would say to make that change at that time.

We're not burning this code onto a CD. We can refactor everything in minutes with a PR. Best to wait for complexity to be required before requiring reviewers to understand and maintain that complexity.

Code that doesn't exist doesn't fail, and you're generally not going to get comments about code that hasn't been written

Reply:
I definitely hold a differing opinion on that. We arent burning this code into a CD, but we also aren't working under a microscope. I think a lot of headaching and refactoring can be saved down the road by adding a little more flexibility/complexity now, especially when we can reasonably predict a few meters of road in front of us.